subsys/modbus/modbus_serial.c

```diff
@@ -499,6 +530,55 @@ int modbus_serial_tx_adu(struct modbus_context *ctx)
 	return -ENOTSUP;
 }
 
+#if defined CONFIG_MODBUS_SERIAL_ASYNC
+static void async_callback(const struct device *dev,
+				   struct uart_event *evt, void *user_data)
+{
+	struct modbus_context *ctx = (struct modbus_context *)user_data;
+	struct modbus_serial_config *cfg;
+
+	if (ctx == NULL) {
+		LOG_ERR("Modbus hardware is not properly initialized");
+		return;
+	}
+	cfg = ctx->cfg;
+
+	switch (evt->type) {
+	case UART_TX_DONE:
+		LOG_DBG("UART_TX_DONE: sent %d bytes", evt->data.tx.len);
+		cfg->uart_buf_ptr = &cfg->uart_buf[0];
+		modbus_serial_tx_off(ctx);
+		modbus_serial_rx_on(ctx);
+		break;
+	case UART_TX_ABORTED:
+		LOG_DBG("Tx aborted");
+		break;
+	case UART_RX_RDY:
+		cfg->uart_buf_ctr += evt->data.rx.len;
+		k_work_submit(&ctx->server_work);
```

Comment:
How does it handle ASCII mode?

Reply:
mmh, at the current state of this PR, ASCII is not supported in async mode. 
I guess it could be added using similar if statements as in cb_handler_rx and cb_handler_tx inside of the async callback cases ? 
What is your view ?

Reply:
I do not see any benefit to use ASYNC API for ASCII mode. I think if interface enables ASCII mode then modbus serial support should fallback to IRQ API.